### PR TITLE
p2p: remove dial sleep and provide disconnect cooldown

### PR DESCRIPTION
**Pull request by tychoish**
Alternative proposal for #8826

### CODE REVIEW

File: internal/p2p/peermanager.go [hunk 1]
```diff
@@ -144,6 +144,10 @@ type PeerManagerOptions struct {
 	// retry times, to avoid thundering herds. 0 disables jitter.
 	RetryTimeJitter time.Duration
 
+	// DisconnectCooldownPeriod is the amount of time after we
+	// disconnect from a peer before we'll consider dialing a new peer
+	DisconnectCollodown time.Duration
```
Comment (JayT106) [inline internal/p2p/peermanager.go:149]: typo: DisconnectCollodown

**Comment by williambanfield** [comment]
@tychoish mind elaborating what the failure looks like?

> "node after being killed fails to get connected to the network"

**Comment by tychoish** [comment]
> @tychoish mind elaborating what the failure looks like?
> 
> > "node after being killed fails to get connected to the network"

It's a pretty short error log: 
https://github.com/tendermint/tendermint/runs/7005956946?check_suite_focus=true

I'm not sure what's going on, but it looks like after a pertubation a node doesn't catch up and make progress. I've added some more info to the error message which is good in general and might help us see where things get stuck.

**Comment by tychoish** [comment]
I believe that the e2e failure has been resolved.

File: internal/p2p/peermanager.go [hunk 2]
```diff
@@ -867,6 +875,11 @@ func (m *PeerManager) Disconnected(ctx context.Context, peerID types.NodeID) {
 	delete(m.evicting, peerID)
 	delete(m.ready, peerID)
 
+	if peer, ok := m.store.Get(peerID); ok {
+		peer.LastDisconnected = time.Now()
```
Comment (williambanfield) [inline internal/p2p/peermanager.go:879]: We'll need to start a routine that calls the dial waker after the cooldown. See the dial failed code as an example.

File: internal/p2p/peermanager.go [hunk 3]
```diff
@@ -867,6 +875,22 @@ func (m *PeerManager) Disconnected(ctx context.Context, peerID types.NodeID) {
...
+		_ = m.store.Set(peer)
+		// launch a thread to ping the dialWaker when the
+		// disconnected peer can be dialed again.
+		go func() {
+			timer := time.NewTimer(time.Until(time.Now().Add(m.options.DisconnectCooldownPeriod)))
```
Comment (williambanfield) [inline internal/p2p/peermanager.go:884]: Can this not just be the duration directly? Any reason to do the until...add stuff?

File: proto/tendermint/p2p/types.proto [hunk 4]
```diff
@@ -29,10 +29,12 @@ message NodeInfoOther {
...
 message PeerInfo {
-  string                    id             = 1 [(gogoproto.customname) = "ID"];
-  repeated PeerAddressInfo  address_info   = 2;
-  google.protobuf.Timestamp last_connected = 3 [(gogoproto.stdtime) = true];
-  bool                      inactive       = 4;
+  string                    id                = 1 [(gogoproto.customname) = "ID"];
+  repeated PeerAddressInfo  address_info      = 2;
+  google.protobuf.Timestamp last_connected    = 3 [(gogoproto.stdtime) = true];
+  bool                      inactive          = 4;
+  google.protobuf.Timestamp last_disconnected = 5 [(gogoproto.stdtime) = true];
```
Comment (williambanfield) [inline proto/tendermint/p2p/types.proto:36]: The cooldown is going to be order of seconds/milliseconds, I think it's not that meaningful to persist it.

**Review by williambanfield — COMMENTED** [review]
Wake looks good. A nit and suggestion to not persist the last disconnect.

File: internal/p2p/peermanager.go (hunk 3 above)
Comment (tychoish) [reply internal/p2p/peermanager.go:884]: oh, good call

File: proto/tendermint/p2p/types.proto (hunk 4 above)
Comment (williambanfield) [reply proto/tendermint/p2p/types.proto:36]: Connection throttle is a few ms. You'd have to crash and recover in an instant for that to be hit?

Comment (tychoish) [reply proto/tendermint/p2p/types.proto:36]: ok ok

**Review by williambanfield — APPROVED** [review]

**Pull request merged**